File: lim/전처리/전처리 - 최소 에너지 비용.py
'''
i자리까지의 최소 시세로 계산한다
'''
import sys


def sol(n, fees, prices):
    min_price_at = list(prices)
    for i in range(1, n):
        min_price_at[i] = min(min_price_at[i], min_price_at[i-1])

    total = 0
    for i in range(n-1):
        total += fees[i] * min_price_at[i]

    print(total)


n = int(sys.stdin.readline())
fees = list(map(int, sys.stdin.readline().split()))
prices = list(map(int, sys.stdin.readline().split()))

sol(n, fees, prices)

File: lim/전처리/전처리_최소_에너지_비용.java
package lim.전처리;

import java.util.*;
import java.io.*;

public class 전처리_최소_에너지_비용 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int n;
    static int[] costs, prices;

    static void sol() throws IOException {
        long minPrice = prices[0];
        long answer = minPrice * costs[0];
        for (int i = 1; i < costs.length; i++) {
            minPrice = Math.min(minPrice, prices[i]);
            answer += minPrice * costs[i];
        }
        bw.write(answer + "");
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());

        costs = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n - 1; i++) {
            costs[i] = Integer.parseInt(st.nextToken());
        }

        prices = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            prices[i] = Integer.parseInt(st.nextToken());
        }
        sol();
        br.close();
        bw.flush();
        bw.close();
    }
}
